### PR TITLE
add priority to the creating process of the task

### PR DESCRIPTION
**Pull request by titoOdUA**
exploring solutions for the https://github.com/alanvardy/tod/issues/336

### CODE REVIEW

**Comment by titoOdUA** [comment]
Hi, @alanvardy 
I have created pull request to explore possible solutions
One thing that really bugs me is priorities

Todoist's REST API have priorities as 1,2,3,4. 4 being the highest.
Todoist's SYNC API have priorities as p1, p2, p3, p4. p1 being the highest......

Actually there was already the ability to add priority to the task via `tod task create` by providing it in the text (as p1. p2 ...etc.). As I understand the problem was that `tod task create` asking for the priority via prompt and you can't use it in alfred/rofi. 

Maybe it would be easier to just get reed of the prompt and just add to the help text that you can specify a priority there, as well as a bunch of things (project, assignee, date, label, description) 

There is a possibility to just use a SYNC api for all operations. Parse the string with text, convert it to Item struct value, and from there just pass it around to all the methods that need it.

But that's a sizable refactor of course =)

In regards to the current issue to add a priority flag, how do you like the solution in the pull request?

File: src/main.rs [hunk 1]
```diff
@@ -279,6 +281,16 @@ fn project_schedule(matches: &ArgMatches) -> Result<String, String> {
...
 }
 
+#[cfg(not(tarpaulin_include))]
+fn priority_arg() -> Arg {
+    Arg::new("priority")
+        .long("priority")
+        .num_args(1)
+        .required(false)
+        .value_name("PRIORITY")
+        .help("Priority from 1(without priority) to 4(highest)")
```
Comment (alanvardy) [inline src/main.rs:291]:
```suggestion
        .help("Priority from 1 (without priority) to 4 (highest)")
```

**Review by alanvardy — APPROVED** [review]

File: src/main.rs [hunk 2]
```diff
@@ -181,16 +182,17 @@ fn fetch_project(matches: &ArgMatches, config: &Config) -> Result<String, String
 fn task_create(matches: &ArgMatches) -> Result<String, String> {
     let config = fetch_config(matches)?;
     let content = fetch_string(matches, "content", "Content")?;
+    let priority = fetch_string(matches, "priority", "Priority")?;
```
Comment (alanvardy) [inline src/main.rs:185]: Actually, can we make this a select? As there are only 4 valid options.

**Comment by alanvardy** [comment]
> Hi, @alanvardy 
> I have created pull request to explore possible solutions
> One thing that really bugs me is priorities
> 
> Todoist's REST API have priorities as 1,2,3,4. 4 being the highest.
> Todoist's SYNC API have priorities as p1, p2, p3, p4. p1 being the highest......
> 
> Actually there was already the ability to add priority to the task via `tod task create` by providing it in the text (as p1. p2 ...etc.). As I understand the problem was that `tod task create` asking for the priority via prompt and you can't use it in alfred/rofi. 
> 
> Maybe it would be easier to just get reed of the prompt and just add to the help text that you can specify a priority there, as well as a bunch of things (project, assignee, date, label, description) 
> 
> There is a possibility to just use a SYNC api for all operations. Parse the string with text, convert it to Item struct value, and from there just pass it around to all the methods that need it.
> 
> But that's a sizable refactor of course =)
> 
> In regards to the current issue to add a priority flag, how do you like the solution in the pull request?

That's right. I would prefer to use the REST API as it is a little less weird for me, but the SYNC API is just so useful for the quick add. 

I guess I like the option of having both. The quick add API could use a little extra documentation letting people know the options available, but the regular task create holds people by the hand a bit more

File: src/projects.rs [hunk 3]
```diff
@@ -263,8 +263,12 @@ pub fn add_item_to_project(
     config: &Config,
     content: String,
     project: &str,
+    priority: String,
```
Comment (alanvardy) [inline src/projects.rs:266]: I think that it would be better if we pass the enum around rather than the string, thoughts?

**Review by alanvardy — CHANGES_REQUESTED** [review]
Sorry, I missed a couple of things in the first review.

Would like a `select` menu for the priority when the flag is not passed in, for the people who are not familiar with the API, and think that the `add_item_to_project` would do better with the enum passed in

File: src/main.rs (hunk 2 above)
Comment (titoOdUA) [reply src/main.rs]: I don't really sure that I have understood you correctly.
I made some changes, can you tell is it what you had in mind?

Basically if flag wasn't provided - there is prompt for priority now when you use `task create`
I will clean code in next commit (maybe encapsulate some logic in Priority struct, create separate file for it etc)

File: src/projects.rs (hunk 3 above)
Comment (titoOdUA) [reply src/projects.rs]: yeah, missed it, thanks

File: src/main.rs (hunk 2 above)
Comment (alanvardy) [reply src/main.rs]: Yeah, absolutely. The modification I am hoping for is to use [Config::select_input)[https://github.com/alanvardy/tod/blob/5b2fd5b060510d6a1786d33495a7beb035569052/src/config.rs#L246] instead of `Config::get_input` so that the user can select the priority (but can still just type the number too)

Comment (titoOdUA) [reply src/main.rs]: As I understand I have basically the same thing right now

```rust
#[cfg(not(tarpaulin_include))]
fn task_create(matches: &ArgMatches) -> Result<String, String> {
    use inquire::Select;
    use items::Priority;

    let config = fetch_config(matches)?;
    let content = fetch_string(matches, "content", "Content")?;
    let priority = match Priority::get_from_matches(&matches) {
        Some(value) => value,
        None => {
            let options = vec![
                Priority::None,
                Priority::Low,
                Priority::Medium,
                Priority::High,
            ];
            Select::new(
                "Choose a priority that should be assigned to task:",
                options,
            )
            .prompt()
            .map_err(|e| e.to_string())
            .expect("Failed to create option list of priorities")
        }
    };

    let project = fetch_project(matches, &config)?;

    projects::add_item_to_project(&config, content, &project, priority)
}

```

you want me to reuse Config::select_input in this situation to avoid dublication? But I can't give it a vec of Priority, only a vec of strings.
I can make it generic so it can take a type that implements Display, like this

```rust
pub fn select_input<T: Display>(desc: &str, options: Vec<T>) -> Result<T, String> {
    Select::new(desc, options)
        .prompt()
        .map_err(|e| e.to_string())
}

```

but the part with test data must go or it will not compile (at least I don't know how to make it compile with it)
```rust
    if cfg!(test) {
        return Ok(String::from("Africa/Asmera"));
    }
```

Maybe I missing something =)

Comment (alanvardy) [reply src/main.rs]: I like your implementation with the generic type that implements display. What if we got the select to return the first item of the Vector when `config!(test)` That way the function would always return the correct type.

Comment (titoOdUA) [reply src/main.rs]: Hi,
I have commited your suggestion. But now there are problems with config_tests tests because of the fact that the method returns different timezone (first from the list).
I fix one test and another becomes broken =) They are hard to follow. There is need for some juggling with data.
 
 I would be grateful if you can help me to resolve them =)

As I understand `if cfg!(test)` there for "mocking the user input"?
It looks somewhat messy and hard to follow in general, wonder if there is a better or simpler way to do it.

**Comment by alanvardy** [comment]
Current dependencies on/for this PR:
* main
  * **PR #347** <a href="https://app.graphite.dev/github/pr/alanvardy/tod/347" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/alanvardy/tod/347?utm_source=stack-comment).

File: src/main.rs (hunk 2 above)
Comment (alanvardy) [reply src/main.rs]: I pulled in your branch and pushed it up to this repo on this PR here: https://github.com/alanvardy/tod/pull/351

I agree that there's lots of improvement needed around the testing, I'm kinda figuring it out as I go and don't have a solid grasp on what standard practice is in the Rust community :+1: 

That's right, I'm just wanting to mock user input, because without it the test would just fail on CI

Closing this PR here so we can continue with the new one.
